docs/proposals/runtime-client-server.md

```diff
@@ -0,0 +1,143 @@
+
+However, this enables container runtime implementations to be built and maintained outside of the core kubernetes tree.
+
+## Proposed design
+
+**Design of client/server container runtime**
+
+The main idea of client/server container runtime is to keep main control logic in kubelet while letting remote runtime only do dedicated actions:
+
+* Keep `SyncPod()` in kubelet and remote runtime only needs to start or kill a pod
+* Keep volume manager in kubelet and pass volumeMap to remote runtime when starting a new pod
+* Keep image garbage collection in kubelet and remote runtime only needs to pull or remove images
+
+Furthermore, kubelet network plugin and cAdvisor should make sense of container runtime type:
+
+* Kubelet network plugin has no sense of dedicated container runtime currently, but new container runtime may not work with existing docker or rkt plugin. The type of container runtime will be added to kubelet network plugin interface.
+* cAdvisor only supports docker and it is not working with other container runtimes. The type of container runtime will be added to cAdvisor Interface. Each new container runtime should add their own driver to cAdvisor.
+
+**`RemoteRuntime` Interface**
+
+```go
+type RemoteRuntime interface {
+  Type() string
+  Version() (Version, error)
+
+  // Pod APIs
+  GetPods(all bool) ([]*Pod, error)
+  GarbageCollect(gcPolicy ContainerGCPolicy) error
+  KillPod(pod *api.Pod, runningPod Pod) error
+  RunPod(pod *api.Pod, volumes kubecontainer.VolumeMap, options []kubecontainer.RunContainerOptions) error
+  GetPodStatus(*api.Pod) (*api.PodStatus, error)
+
+  // Image APIs
+  PullImage(image ImageSpec, pullSecrets []api.Secret) error
+  IsImagePresent(image ImageSpec) (bool, error)
+  ListImages() ([]Image, error)
+  RemoveImage(image ImageSpec) error
+
+  // Container APIs
+  GetContainerLogs(pod *api.Pod, containerID ContainerID, logOptions *api.PodLogOptions, stdout, stderr io.Writer) (err error)
+  ContainerCommandRunner
+  ContainerAttacher
+}
+
+type ContainerAttacher interface {
+  AttachContainer(id ContainerID, stdin io.Reader, stdout, stderr io.WriteCloser, tty bool) (err error)
+}
+
+type ContainerCommandRunner interface {
+  RunInContainer(containerID ContainerID, cmd []string) ([]byte, error)
+  ExecInContainer(containerID ContainerID, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser, tty bool) error
+  PortForward(pod *Pod, port uint16, stream io.ReadWriteCloser) error
+}
+```
+
+**Kubelet network plugin interface**
+
+```go
+// Plugin is an interface to network plugins for the kubelet
+type NetworkPlugin interface {
+	// Init initializes the plugin.  This will be called exactly once
+	// before any other methods are called.
+	Init(host Host, containerRuntime string) error
+
+	// Name returns the plugin's name. This will be used when searching
+	// for a plugin by name, e.g.
+	Name() string
+
+	// SetUpPod is the method called after the infra container of
```

Comment:
@yifan-gu Thanks for figuring out. There no concept of infra container in hyper. I will modify this comments to pod specific.